library/alloc/src/boxed.rs

```diff
@@ -1231,9 +1232,20 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-unsafe impl<#[may_dangle] T: ?Sized, A: Allocator> Drop for Box<T, A> {
+#[rustc_const_unstable(feature = "const_box", issue = "92521")]
+unsafe impl<#[may_dangle] T: ?Sized, A: ~const Allocator> const Drop for Box<T, A> {
+    #[inline]
     fn drop(&mut self) {
-        // FIXME: Do nothing, drop is currently performed by compiler.
+        // the T in the Box is dropped by the compiler before the destructor is run
```

Comment:
Btw, you've probably seen this before, but just to bring everyone up to speed, one way this can be made less magical is an `unsafe trait DerefMove` (with similar "must be uniformly `impl`'d regardless of generic parameters" restriction as `Drop`) that makes a type into "split content and deallocation".

`DerefMove::deref_move`, without new language features, would likely need to return `-> *mut T`, and the compiler would use it wherever it assumes today built-in deref works on `Box<T>` (and `Deref`/`DerefMut` would be too weak), *including* the `Drop(*some_box)` it does just before invoking `Drop::drop`.

Reply:
This idea is actually the end goal of all my Box related PRs